### PR TITLE
Mirror codebase structure in tests

### PR DESCRIPTION
**Pull request by Armavica**
**What is this PR about?**

This PR splits up the test files `test_distribution.py`, `test_moments.py` and `test_random.py`, to start tackling issue #5777.

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes
- ...

## Bugfixes / New features
- ...

## Docs / Maintenance
- Break up `test_distribution.py`, `test_moments.py` and `test_random.py` in more specific files.

### CODE REVIEW

File: pymc/tests/distributions/test_distribution.py [hunk 1]
```diff
@@ -0,0 +1,499 @@
+#   Copyright 2020 The PyMC Developers
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py:1]: This one seems to have logic that makes more sense in other test files, I will leave specific suggestions sometime this week

Comment (ricardoV94) [reply pymc/tests/distributions/test_distribution.py:1]: test_random and test_moments might be better of split across test_discrete/continous/multivariate/...

Comment (Armavica) [reply pymc/tests/distributions/test_distribution.py:1]: Yes, I was indeed thinking to do random and moments next.

**Comment by ricardoV94** [comment]
> This draft PR is a WIP to tackle issue #5777, posted early to avoid duplication of effort.

To avoid conflicts with the other PRs I would suggest we merge this earlier rather than later.

File: pymc/tests/distributions/test_distribution.py [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+            lambda _: domain.lower[1],
+            lambda _: domain.upper[1],
+            lambda _, __: domain.lower[2],
+            lambda _, __: domain.upper[2],
+        )[0]
+    else:
+        raise ValueError("Dont know how to integrate shape: " + str(shape))
+
+
+def test_hierarchical_logp():
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_logprob`

File: pymc/tests/distributions/test_distribution.py [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+
+    logp_ancestors = list(ancestors([m.logp()]))
+    ops = {a.owner.op for a in logp_ancestors if a.owner}
+    assert len(ops) > 0
+    assert not any(isinstance(o, RandomVariable) for o in ops)
+    assert x.tag.value_var in logp_ancestors
+    assert y.tag.value_var in logp_ancestors
+
+
+def test_hierarchical_obs_logp():
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_logprob`

File: pymc/tests/distributions/test_distribution.py [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+        pm.Uniform("y", x, 2, observed=obs)
+
+    logp_ancestors = list(ancestors([model.logp()]))
+    ops = {a.owner.op for a in logp_ancestors if a.owner}
+    assert len(ops) > 0
+    assert not any(isinstance(o, RandomVariable) for o in ops)
+
+
+class TestMatchesScipy:
+    def test_truncated_normal(self):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_continuous`

File: pymc/tests/distributions/test_distribution.py [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+        check_logp(
+            TruncatedNormal,
+            R,
+            {"mu": R, "sigma": Rplusbig, "lower": -Rplusbig},
+            functools.partial(scipy_logp, upper=np.inf),
+            decimal=select_by_precision(float64=6, float32=1),
+            skip_paramdomain_outside_edge_test=True,
+        )
+
+    def test_get_tau_sigma(self):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_continuous` or `dist_math` (wherever it is implemented)

File: pymc/tests/distributions/test_distribution.py [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+            (7.5, 2.000, 5.000, 5.000, -2.8259429),
+            (15.0, 5.000, 7.500, 7.500, -3.3093854),
+            (50.0, 50.000, 10.000, 10.000, -3.6436067),
+            (1000.0, 500.000, 10.000, 20.000, -27.8707323),
+            (-1.0, 1.0, 20.0, 0.9, -3.91967108),  # Fails in scipy version
+            (0.01, 0.01, 100.0, 0.01, -5.5241087),  # Fails in scipy version
+            (-1.0, 0.0, 0.1, 0.1, -51.022349),  # Fails in previous pymc version
+        ],
+    )
+    def test_ex_gaussian(self, value, mu, sigma, nu, logp):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_continuous`

File: pymc/tests/distributions/test_distribution.py [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+            ExGaussian("eg", mu=mu, sigma=sigma, nu=nu)
+        pt = {"eg": value}
+        assert_almost_equal(
+            model.compile_logp()(pt),
+            logp,
+            decimal=select_by_precision(float64=6, float32=2),
+            err_msg=str(pt),
+        )
+
+    def test_gaussianrandomwalk(self):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_timeseries`

File: pymc/tests/distributions/test_distribution.py [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+            Binomial("a", n=5, p=0.5, transform="log")
+        err.match("Transformations for discrete distributions")
+
+
+@pytest.mark.parametrize(
+    "sparse, size",
+    [(False, ()), (False, (1,)), (False, (4,)), (False, (4, 4, 4)), (True, ()), (True, (4,))],
+    ids=str,
+)
+def test_car_logp(sparse, size):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_multivariate`

File: pymc/tests/distributions/test_distribution.py [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+        tol = 1e-5
+    assert np.allclose(delta_logp - delta_logp[0], 0.0, atol=tol)
+
+
+@pytest.mark.parametrize(
+    "sparse",
+    [False, True],
+    ids=str,
+)
+def test_car_matrix_check(sparse):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_multivariate`

File: pymc/tests/distributions/test_distribution.py [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", ".*number of samples.*", UserWarning)
+            pm.sample(draws=5, tune=1, mp_ctx="spawn")
+
+    import cloudpickle
+
+    cloudpickle.loads(cloudpickle.dumps(y))
+
+
+def test_distinct_rvs():
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be removed (separate commit) or placed in `test_sampling`?

File: pymc/tests/distributions/test_distribution.py [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+    a_val = npr.normal(loc=mu_val, scale=1, size=to_tuple(size) + (supp_shape,)).astype(
+        aesara.config.floatX
+    )
+    log_densityt = joint_logp(a, a.tag.value_var, sum=False)[0]
+    assert log_densityt.eval(
+        {a.tag.value_var: a_val, mu.tag.value_var: mu_val},
+    ).shape == to_tuple(size)
+
+
+class TestLKJCholeskCov:
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_multivariate`

File: pymc/tests/distributions/test_distribution.py [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+        x = pm.LKJCholeskyCov.dist(
+            n=3, eta=1, sd_dist=sd_dist, size=size, shape=shape, compute_corr=False
+        )
+        resized_sd_dist = x.owner.inputs[-1]
+        assert resized_sd_dist.eval().shape == (10, 3)
+        # LKJCov has support shape `(n * (n+1)) // 2`
+        assert x.eval().shape == (10, 6)
+
+
+def test_constantdist_deprecated():
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This should be in `test_discrete`

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

File: pymc/tests/distributions/test_distribution.py [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+)
+from pymc.distributions.continuous import get_tau_sigma
+from pymc.distributions.shape_utils import to_tuple
+from pymc.model import Model
+from pymc.tests.distributions.util import Nat, R, Rplus, Rplusbig, Vector, check_logp
+from pymc.tests.helpers import select_by_precision
+from pymc.vartypes import continuous_types
+
+
+def integrate_nd(f, domain, shape, dtype):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: I think we are not using this helper anymore, so we can remove it (separate commit) or move it to `util`

**Comment by Armavica** [comment]
> To avoid conflicts with the other PRs I would suggest we merge this earlier rather than later.

Thank you. I don't mind rebasing, but I see what I can do tonight.

**Comment by ricardoV94** [comment]
I didn't mean you have to rush like that. Just giving it priority over other PRs.

**Comment by Armavica** [comment]
The files `test_distribution.py`, `test_moment.py` and `test_random.py` are more or less done. I am just not sure what to do with the final test that remains in `test_random.py`.

I am planning to work on the others this weekend (`gp`, `ode`, `smc`, `step_methods`, `hmc`, `tuning`, `variational`), but you can merge this one now if you want, it is fine for me either way.

File: pymc/tests/distributions/test_random.py [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+
+import numpy as np
+import pytest
+
+import pymc as pm
+
+from pymc.tests.helpers import SeededTest
+
+
+class TestNestedRandom(SeededTest):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_random.py]: I think we can put this in `test_sampling`?

File: pymc/tests/distributions/test_timeseries.py [hunk 15]
```diff
@@ -36,9 +37,50 @@
...
+    R,
+    Rplus,
+    Vector,
+    assert_moment_is_expected,
+    check_logp,
+)
+from pymc.tests.helpers import SeededTest, select_by_precision
+
+
+class TestMatchesScipy:
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_timeseries.py]: This is repeated.  Also you can simply put the `test_logp` inside the TestGaussianRandomWalk class, not need for the MatchesScipy class. Just call it `test_logp` or something like that

File: pymc/tests/distributions/test_distribution.py [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+    unexpected_not_implemented = missing_moments - not_implemented
+    if unexpected_not_implemented:
+        raise NotImplementedError(
+            f"Unexpected by this test, distributions {unexpected_not_implemented} do "
+            "not have a `moment` implementation. Either add a moment or filter "
+            "these distributions in this test."
+        )
+
+
+class TestDensityDist:
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: Can we put all the other DensityDist tests in this file under this class?

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

**Comment by ricardoV94** [comment]
> I am planning to work on the others this weekend (gp, ode, smc, hmc, tuning, variational), but you can merge this one now if you want, it is fine for me either way.

Yes let's keep this PR just about the distribution tests

**Comment by ricardoV94** [comment]
Looks nearly done. I left some final suggestions. Once you address them I think we can merge.

**Comment by ricardoV94** [comment]
Can you add a bullet point in the maintenance section on the top PR post?

Also some tests seem to be failing, let me know if you need help with those (haven't peaked at the logs)

**Comment by Armavica** [comment]
Thank you for your review! Yes, the tests should be fixable. The only one that I don't understand is the docs: it fails to build the 404 page for an obscure reason... Do you have an idea about this? And it was the same error for the first version of the PR.

**Comment by ricardoV94** [comment]
I think the docs build has been failing everywhere, you can ignore it

**Comment by Armavica** [comment]
I think it is all set now. We should probably think of rebalancing the github test workers at some point, with many tests having moved around.

**Comment by ricardoV94** [comment]
> We should probably think of rebalancing the github test workers at some point, with many tests having moved around.

Yes, after a few more PRs get open/merged we can get an idea of the average runtime to rebalance.

File: pymc/tests/distributions/test_distribution.py [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+            assert np.all(evaled_moment == 0)
+        else:
+            with pytest.raises(
+                TypeError,
+                match="Cannot safely infer the size of a multivariate random variable's moment.",
+            ):
+                evaled_moment = moment(a).eval({mu: mu_val})
+
+
+def test_all_distributions_have_moments():
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: This seems to be repeated above

File: pymc/tests/distributions/test_distribution.py [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+    unexpected_not_implemented = missing_moments - not_implemented
+    if unexpected_not_implemented:
+        raise NotImplementedError(
+            f"Unexpected by this test, distributions {unexpected_not_implemented} do "
+            "not have a `moment` implementation. Either add a moment or filter "
+            "these distributions in this test."
+        )
+
+
+class TestDensityDist:
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_distribution.py]: There's already a TestDensityDist class above

File: pymc/tests/distributions/test_timeseries.py [hunk 19]
```diff
@@ -36,9 +37,50 @@
...
+            pm.GaussianRandomWalk,
+            Vector(R, 4),
+            {"mu": R, "sigma": Rplus, "steps": Nat},
+            ref_logp,
+            decimal=select_by_precision(float64=6, float32=1),
+            extra_args={"init_dist": Normal.dist(0, 100)},
+        )
+
+
+class TestMatchesScipy:
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_timeseries.py]: This test is repeated above and below

**Review by ricardoV94 — CHANGES_REQUESTED** [review]

**Comment by Armavica** [comment]
Argh, my bad, tricky rebasing… Good catches.

**Comment by canyon289** [comment]
@Armavica youre doing an amazing job. Thanks for all the effort

File: pymc/tests/distributions/test_continuous.py [hunk 20]
```diff
@@ -0,0 +1,2229 @@
...
+        (
+            np.arange(1, 6),
+            np.arange(1, 6),
+            np.arange(1, 6),
+            (2, 5),
+            np.full((2, 5), (1.0, 1.25, 2.111111111111111, 3.0625, 4.04)),
+        ),
+    ],
+)
+def test_asymmetriclaplace_moment(b, kappa, mu, size, expected):
```
Comment (ricardoV94) [inline pymc/tests/distributions/test_continuous.py]: This one is missing an indentation, which is causing the following tests to be inside it and not run (codecov warning)

File: pymc/tests/distributions/test_distribution.py (hunk 17 above)
Comment (ricardoV94) [reply pymc/tests/distributions/test_distribution.py]: Still looks unresolved on my end? Codecov is complaining the test above are not being covered, which agress with them still being duplicated.

File: pymc/tests/distributions/test_distribution.py (hunk 18 above)
Comment (ricardoV94) [reply pymc/tests/distributions/test_distribution.py]: Still duplicated above?

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Still some duplicated tests in `test_distributions` and an indentation error that is putting two tests inside the previous one. I know how painful rebases can be :D

Otherwise looks great!!!

**Comment by Armavica** [comment]
Ok, I understood my mistake: I think I was repeatedly trying to fix the wrong commit. The rebase went without problem this time so I think that it should be ok. And thank you for mentioning codecov: I understood where to look to spot all of this.

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Merged! Thanks for this :)  Feel free to break some more eggs.

**Pull request merged**